GuruLoader/GuruLoader.cs

```
@@ -237,13 +266,49 @@ public static string ComposeGuruUrl(string cik)
+            });
+
+        foreach (var port in ports) {
+            foreach (var pos in port.Positions) {
+                DisplayCompany dp;
+                if (recentCompanies.TryGetValue(FormKeyD(pos), out dp))
+                    dp.ChangePositions.Add(new DisplayChangePosition {
+                        Change = pos.Change, IsNew = pos.IsNew, IsSold = pos.IsSold, Shares = pos.Shares,
+                        Value = pos.Value, Price = pos.Price, PercOfPortfolio = pos.PercOfPortfolio,
+                        Date = port.EndQuarterDate
+                    });
+            }
+        }
+        return recentCompanies.Values;
+    }
+
     // Utility function to help waiting for all portfolios to be loaded
     async static Task<Portfolio> FetchPortfolio(HttpClient client, string htmlLink) {
-        using (var html = await client.GetStreamAsync(htmlLink)) {
-            var submissionLink = ParseHtmFile(html);
-            using (var submissionStream = await client.GetStreamAsync(MakeSecLinkAbsolute(submissionLink))) {
-                return ParseSubmissionFile(submissionStream);
+        try {
+            using (var html = await client.GetStreamAsync(htmlLink)) {
+                var submissionLink = ParseHtmFile(html);
+                using (var submissionStream = await client.GetStreamAsync(MakeSecLinkAbsolute(submissionLink))) {
+                    return ParseSubmissionFile(submissionStream);
+                }
             }
+        } catch(Exception e) {
+            // Before a certain date (2013) the format for 13F was different, not XML, hence the code will throw
+            // a different exception depending on the particular format of the text file read.
+            // It is unclear if it's of any interest to look so far back in the past.
+            // Until I implement support for such old files, I construct an empty portfolio in such cases
+            // TODO: fix the code so that doesn't swallow exceptions
```

To guard against swallowing too many exceptions, it might be advisable to catch only the types of exceptions you expect to be thrown (some sort of IO exception, most likely). That way, if something else is thrown it won't be silently swallowed.

---

This is tricky. Most of the time I get a ArgumentOutOfRangeException because the text parsing of the file fails, but I could get something else. The real fix is to do some more intelligent parsing of different kinds of files that I could get before a certain date. But that is of little interest to the user and perhaps not all that valuable for a demo app. But it is troublesome, hence the TODO in the code. I also added a line to be notified if I see a different kind of exception.